operator/pkg/ciliumidentity/reconciler.go

```diff
@@ -0,0 +1,438 @@
+	podStore, err := pod.Store(ctx)
+	if err != nil {
+		return nil, err
+	}
+	cidStore, err := ciliumIdentity.Store(ctx)
+	if err != nil {
+		return nil, err
+	}
+	cepStore, err := ciliumEndpoint.Store(ctx)
+	if err != nil {
+		return nil, err
+	}
+	cesStore, err := ciliumEndpointSlice.Store(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	r := &reconciler{
+		logger:          logger,
+		ctx:             ctx,
+		clientset:       clientset,
+		idAllocator:     idAllocator,
+		desiredCIDState: NewCIDState(logger),
+		cidUsageInPods:  NewCIDUsageInPods(),
+		cidUsageInCES:   NewCIDUsageInCES(),
+		queueOps:        queueOps,
+		nsStore:         nsStore,
+		podStore:        podStore,
+		cidStore:        cidStore,
+		cepStore:        cepStore,
+		cesStore:        cesStore,
+		cesEnabled:      cesEnabled,
+	}
+
+	return r, nil
+}
+
+// syncCESsOnStartup updates the cache of CID usage in CES for all the
+// existing CESs.
+func (r *reconciler) calcDesiredStateOnStartup() error {
+	r.syncCESsOnStartup()
+	return r.syncPodsOnStartup()
+}
+
+func (r *reconciler) syncCESsOnStartup() {
+	if !r.cesEnabled {
+		return
+	}
+
+	for _, ces := range r.cesStore.List() {
+		r.cidUsageInCES.ProcessCESUpsert(ces.Name, ces.Endpoints)
+	}
+}
+
+// syncPodsOnStartup ensures that all pods have a CID for their labels.
+func (r *reconciler) syncPodsOnStartup() error {
+	var lastError error
+
+	for _, pod := range r.podStore.List() {
+		if err := r.reconcilePod(podResourceKey(pod.Name, pod.Namespace)); err != nil {
+			lastError = err
+		}
+	}
+
+	return lastError
+}
+
+// reconcileCID ensures that the desired state for the CID is reached, by
+// comparing the CID in desired state cache and watcher's store and doing one of
+// the following:
+// 1. Nothing - If CID doesn't exist in both desired state cache and watcher's
+// store.
+// 2. Creates CID - If CID only exists in the desired state cache.
+// 3. Updates CID - If CIDs in the desired state cache and watcher's store are
+// not the same.
+// Currently, the CID deletion is handled by the operator/identitygc
+func (r *reconciler) reconcileCID(cidResourceKey resource.Key) error {
+	cidName := cidResourceKey.Name
+	storeCID, existsInStore, err := r.cidStore.GetByKey(cidResourceKey)
+
+	if err != nil && !k8serrors.IsNotFound(err) {
+		return err
+	}
+
+	cidKey, existsInDesiredState := r.desiredCIDState.LookupByID(cidName)
+	if !existsInDesiredState && !existsInStore {
+		err := r.makeIDAvailable(cidName)
+		r.logger.Warn("Failed to return CID to pool",
+			logfields.CIDName, cidName,
+			logfields.Error, err)
+		return nil
+	}
+
+	cidIsUsed := r.cidIsUsedInPods(cidName) || r.cidIsUsedInCEPOrCES(cidName)
+
+	if !existsInDesiredState {
+		if cidIsUsed {
+			r.cidCreateLock.Lock()
+			defer r.cidCreateLock.Unlock()
+
+			id, err := r.idAllocator.ValidateIDString(cidName)
+			if err != nil {
+				return err
+			}
+			err = r.idAllocator.Allocate(idpool.ID(id))
+			if err != nil {
+				return err
+			}
+
+		}
+		return nil
+	}
```

Comment:
Nonblocking: using an early return here makes the code a bit nicer

```suggestion
	if !existsInDesiredState {
		if !cidIsUsed {
			return nil
		}

		r.cidCreateLock.Lock()
		defer r.cidCreateLock.Unlock()

		id, err := r.idAllocator.ValidateIDString(cidName)
		if err != nil {
			return err
		}
		return r.idAllocator.Allocate(idpool.ID(id))
	}
```

Reply:
refactored